New Plugin - MeeWee Time Registration

renielbryn opened the pull request:
## Hey, I just made a Pull Request!

We are contributing a new plugin named Meewee Time Registration plugin.  
![image](https://github.com/backstage/community-plugins/assets/58383463/e44dcc5a-f0fc-49cd-ba79-cf1fb71f7798)
![image](https://github.com/backstage/community-plugins/assets/58383463/6eb5af21-ed89-4956-a795-31a88ab4677c)


#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

awanlin commented on .github/CODEOWNERS:
```diff
@@ -27,3 +27,4 @@ yarn.lock                                           @backstage/community-plugins
 /workspaces/sonarqube                               @backstage/community-plugins-maintainers  @backstage/sda-se-reviewers
 /workspaces/user-settings                           @backstage/community-plugins-maintainers  @backstage/sda-se-reviewers
 /workspaces/tech-insights                           @backstage/community-plugins-maintainers  @xantier
+/workspaces/meewee-time-registration                @backstage/community-plugins-maintainers  @dewise-team/default
```
Please make sure to add a newline at the end of the file

renielbryn replied:
New line added 👍

awanlin commented on workspaces/meewee-time-registration/README.md:
```diff
@@ -0,0 +1,10 @@
+# [Backstage](https://backstage.io)
```
Can we update this with a proper title and perhaps links to the installation docs for each plugin?

renielbryn replied:
Done 👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/README.md:
```diff
@@ -0,0 +1,36 @@
+
+This plugin is contributed by [Dewise](https://www.dewise.com/)
+
+## Features
```
Screen shots are really helpful, you have some in the PR description that would be great to add here!

renielbryn replied:
Done 👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/package.json:
```diff
@@ -0,0 +1,61 @@
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "license": "Apache-2.0",
+  "private": true,
```
You'll want to remove this or it won't get published to NPM.

renielbryn replied:
Done 👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/App.tsx:
```diff
@@ -0,0 +1,13 @@
+import { AuthProvider } from './context/AuthContext';
+import MainComponent from './components/MainComponent/MainComponent';
+
+const App: React.FC = () => {
```
`React.FC` is to be avoided: https://backstage.io/docs/architecture-decisions/adrs-adr006

renielbryn replied:
Done 👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/axios/index.ts:
```diff
@@ -0,0 +1,70 @@
+    };
+
+    const response = await client({
+      baseURL: 'https://api.staging.meewee.com/0/',
```
This looks to be hard coded?

renielbryn replied:
Done 👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/axios/index.ts:
```diff
@@ -0,0 +1,70 @@
+
+    const response = await client({
+      baseURL: 'https://api.staging.meewee.com/0/',
+      //signal: isApiGetRegistrations ? controller.signal : undefined,
```
This code can be removed if not needed 👍

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/components/MainComponent/MainComponent.tsx:
```diff
@@ -0,0 +1,209 @@
+    dispatch({ type: 'LOGIN', payload: { userName, token } });
+  };
+
+  // const removeAuthTokenLogout = () => {
```
This can be removed if not needed 👍

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/components/MainComponent/MainComponent.tsx:
```diff
@@ -0,0 +1,209 @@
+  return (
+    <AuthProvider>
+      <Grid container className="main" spacing={3} direction="column">
+        {/* <Button
```
This can be removed if not needed 👍

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/components/ui/LoginModal.tsx:
```diff
@@ -0,0 +1,202 @@
+import { useApi } from '../../request/useApi';
+import { LoginModalProps } from '../../types/components/loginModal';
+
+const LoginModal: React.FC<LoginModalProps> = ({
```
Again we want to avoid `React.FC`

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/components/ui/LoginModal.tsx:
```diff
@@ -0,0 +1,202 @@
+            <Link
+              className={classes.signUpLink}
+              underline="hover"
+              href="https://nxt.staging.meewee.com/signup"
```
Another hard coded ULR?

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/components/ui/ui.style.ts:
```diff
@@ -0,0 +1,200 @@
+
+export const useCustomHeaderStyles = makeStyles(() => ({
+  '@font-face': {
+    fontFamily: 'InterFont',
```
Fonts should be driven by the Backstage theme and should not come from the plugin itself. This create a poor UX as you have potentially very clashing designs.

awanlin replied:
That goes for all cases of this in this file and anywhere else

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/components/ui/ui.style.ts:
```diff
@@ -0,0 +1,200 @@
+  },
+  registerBtn: {
+    color: 'white',
+    backgroundColor: '#219DB6',
```
Colors should not be hard coded and should be coming from the theme. 👍

awanlin replied:
That goes with all cases in this file and anywhere else

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/context/AuthContext.tsx:
```diff
@@ -0,0 +1,45 @@
+  }
+};
+
+const AuthProvider: React.FC = ({ children }) => {
```
Sorry, want to avoid this again, no `React.FC`.

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/request/api/azdevops/AzDevOpsClient.ts:
```diff
@@ -0,0 +1,131 @@
+
+  private async handleResponse(req: request): Promise<any> {
+    // azure devops scope  '499b84ac-1321-427f-aa17-267ca6975798/.default'
+    const token = await this.authApi.getAccessToken(req.scope);
```
There's an existing `azureApiRef` that would be able to get you the tokens needed for this based on how the Adopter has setup their Backstage integration.

renielbryn replied:
We removed this one. Sorry, this wasn't removed on the former commits. This is a feature that we're planning to implement in the future.

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/request/api/azdevops/AzDevOpsClient.ts:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  async getProjects(): Promise<any> {
+    const url = `${await this.discoveryApi.getBaseUrl('ado')}/projects`;
```
What is 'ado'? This is expecting a plugin Id so not sure where this is defined, haven't seen it in this PR?

awanlin replied:
Same with other cases of this in this file

renielbryn replied:
ado is for azure devops feature that we're planning to integrate with the MeeWee plugin. This is a feature that we're planning to implement in the future. Sorry we forgot to remove the codes related for that feature.

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/request/api/transitBuilder/index.ts:
Why is there any empty file here?

renielbryn replied:
Removed

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/request/buildRequestHeader.ts:
```diff
@@ -0,0 +1,18 @@
+// TESTO
+//import { RootState } from '../../../../../packages/app/src/store';
```
This can be removed is not being used

renielbryn replied:
Done👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/axios/index.ts:
```diff
@@ -0,0 +1,70 @@
+// TESTO
+import axios, { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';
```
We want to use a consistent library here and there is an ADR to follow: https://backstage.io/docs/architecture-decisions/adrs-adr013

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/types/authContext.d.ts:
We generally don't follow the pattern of using `*.d.ts` files, going to leave this open to what the other Maintainers think on this though. What ever is decided would apply to all the files like this.

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/src/request/api/azdevops/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { azDevopsApiRef } from '../../../types/azdevops/types';
+export type { AzDevopsApi } from '../../../types/azdevops/types';
```
There are pretty deep paths and I see that throughout the code here. You'll want to follow what is laid out in [ADR004: Module Export Structure](https://backstage.io/docs/architecture-decisions/adrs-adr004)

renielbryn replied:
Removed👍

awanlin commented on workspaces/meewee-time-registration/plugins/plugin-meewee-time-registration/meewee-time-registration/README.md:
```diff
@@ -0,0 +1,36 @@
+3. Register time to your MeeWee account.
+4. View all registered time by dates.
+
+## Setup
```
After looking over the PR there are some big gaps here. There's no mention of needing/using Azure DevOps here. Also, it would probably be helpful to have a link here on how to create a MeeWee account.

renielbryn replied:
Done👍

awanlin reviewed (changes requested):
Hi thanks for the contribution @renielbryn, I've gone over the code and left comments. 

There is one large structural issue with the way the code is placed, currently it's like this:

```text
workspaces
    /meewee-time-registration
        /plugins
            /plugin-meewee-time-registration
                /meewee-time-registration
```

You have an extra folder here that's going to cause issues. I'd remove the `/plugin-meewee-time-registration` folder. 

Also, please look into the failing builds there several items to take on there.

renielbryn commented:
Hi @vinzscam & @awanlin. We've refactored and cleaned up the code. Let me know if all previous comments have been addressed. 

We're happy to contribute to the community plugins. Thank you very much!

awanlin commented:
Hi @renielbryn, sorry, this one fell off my radar last week, I use the Reviews feature in GitHub and it seems if you don't re-trigger the review it won't show up there. I'll try my best to get this reviewed soon but no promises it won't be before Friday.

awanlin commented on workspaces/meewee-time-registration/plugins/meewee-time-registration/src/types/components/customHeader.d.ts:
We don't use the `*.d.ts` convention, these could just be types, this goes for all the files like this.

awanlin reviewed (changes requested):
Had some time so took another look. Can you please rebase?

awanlin commented:
Hi @renielbryn, haven't seen any updates here, will you be picking this up again? Or have you decided to host this yourself, which might allow you to move a bit faster and iterate.

awanlin commented:
Thanks for the update @fp-dewise, I'm going to close this PR then, if you change your mind we can re-open it. 

When your self hosted plugin is shipped feel free to add it to the Plugin Directory, you just submit a PR with an image and YAML file following this documentation: https://backstage.io/docs/plugins/add-to-directory 👍
